File: client/src/model/python/PythonPackage.ts
import {PythonFilter} from "./PythonFilter";
import PythonModule from "./PythonModule";
import PythonDeclaration from "./PythonDeclaration";

export default class PythonPackage extends PythonDeclaration {

    readonly name: string;
    readonly modules: PythonModule[];

    constructor(name: string, modules: PythonModule[] = []) {
        super();

        this.name = name;
        this.modules = modules;

        this.modules.forEach(it => {
            it.containingPackage = this;
        });
    }

    filter(filter: PythonFilter | void): PythonPackage {
        if (!filter) {
            return this;
        }

        return this;

        // return new PythonPackage(
        //     this.name,
        //     modules.filter
        // );
    }

    parent(): null {
        return null;
    }

    children(): PythonModule[] {
        return this.modules;
    }

    toString(): string {
        return `Package "${this.name}"`;
    }
}

File: api-editor/desktop/src/main/kotlin/com/larsreimann/apiEditor/features/ast/model/PythonPackage.kt
package com.larsreimann.apiEditor.features.ast.model

interface PythonPackage : PythonDeclaration {
    override val id: PythonPackageId
    val modules: List<PythonModule>

    override fun children() = sequence {
        yieldAll(modules)
    }

    fun toMutablePythonPackage() = MutablePythonPackage(
        id = id,
        modules = modules.map { it.toMutablePythonModule() },
    )
}

class MutablePythonPackage(
    override val id: PythonPackageId = PythonPackageId(""),
    override var name: String = "",
    modules: List<MutablePythonModule> = emptyList(),
) : MutablePythonDeclaration(), PythonPackage {

    override val modules = MutableContainmentList(modules)
}

typealias PythonPackageId = PythonDeclarationId<PythonPackage>
